Feature: Grant Extension Form

JoeMacl opened the pull request:
### **Description**
Features the new 'grant extension component' that allows staff to grant extensions to students via the 'grant extension' button located in the units page. Currently this component is frontend-only. No API integration has been implemented yet. Backend endpoints will be added in a future update.

**Important:** Currently, there's an issue preventing admins from accessing the units page. For testing purposes, the grant extension form is temporarily accessible to students. This will be restricted to admin users only in future updates.

The form includes the following fields: 

Student: A dropdown selector.

Extension Duration: A slider ranging from 1 to 30 days.

Reason: A required text area input.

Additional Notes: Optional text area.


### To View the Feature

- Sign in as a student (temporarily accessible for testing) and navigate to any enrolled unit, such as Introduction to Programming.
- Locate the 'grant extension' button in any one of the students tasks
- Open the browser console to verify no errors. On submit, you should see logs for the task, received dialog data, submitted payload, and final form values.

![image](https://github.com/user-attachments/assets/a3b46f1a-4c85-4c0c-8b65-27c2f135462f)
![image](https://github.com/user-attachments/assets/edbc937d-835d-45f4-b012-4d3bfce82411)
![image](https://github.com/user-attachments/assets/633119ff-db80-443f-992f-a1d5516a2706)



## Type of change

- [x] New feature (non-breaking change which adds functionality)


# How Has This Been Tested?

- Validated that the range slider updates and reflects the selected value in the UI
- Confirmed responsive behaviour across different view sizes
- Clicking the "Grant Extension" button triggers console.log
- Dropdown renders expected student options

## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [x] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have requested a review on the Pull Request

JoeMacl committed:
feat: add Grant Extension form component

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.html:
```diff
@@ -0,0 +1,64 @@
+        class="w-full border border-gray-300 rounded-md px-4 py-3 text-base focus:ring-2 focus:ring-blue-500 focus:outline-none"
+        required
+      >
+        <option value="" disabled selected>Select a student</option>
```
This didn't seem to be an issue when I tested but it is possible that the 'Select a student' option is still visible after an option is made. My suggestion is to remove the disabled selected attribute and set it dynamically.

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.html:
```diff
@@ -0,0 +1,64 @@
+      ></textarea>
+    </div>
+
+    <!-- Submit -->
```
Might be good to have some kind of error message in the case of a validation issue. For instance, if the student is not selected or the reason is missing- display an error message below the form field.

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.html:
```diff
@@ -0,0 +1,64 @@
+
+    <!-- Notes -->
+    <div>
+      <label for="notes" class="block text-lg font-semibold text-gray-800 mb-2">Additional Notes</label>
```
Since the notes field is optional, it might be nice for it to be displayed as option. For example, Additional Notes(optional) just to make it clearer to users.

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.html:
```diff
@@ -0,0 +1,64 @@
+
+    <!-- Submit -->
+    <div class="pt-4 text-right">
+      <button
```
This is purely up to you but it may be nice to have an indication that the form is being submitted. Maybe disable the button or a loading spinner of some sort during submission.

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.spec.ts:
```diff
@@ -0,0 +1,22 @@
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
```
Since you have form validation, I recommend doing specs to test that the form is not being submitted when it is invalid. (invalid entries or missing fields)

For the optional notes, check that a missing entry will pass validation and not raise an error. 

Test that the form is being initialised properly. It works locally but having specs for all of these might be nice just in case someone updates this form years later and breaks it!

samindiii commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.ts:
```diff
@@ -0,0 +1,34 @@
+import {Component, OnInit} from '@angular/core';
+import {FormGroup, FormBuilder, Validators, ReactiveFormsModule} from '@angular/forms';
+import {CommonModule} from '@angular/common';
+
```
Great job on the formatting of your code as it is very easy to read and understand. Might be useful to add more comments to explain the purpose of certain actions, such as what the logic is or why certain validations are used etc.  This will help future students etc.

samindiii reviewed (changes requested):
Great job on your work! I have reviewed it and have suggested some changes- if you want to discuss them with me, feel free to shoot me a message on Teams! :)

JoeMacl committed:
feat: updated grant extension form with validation and styling

JoeMacl commented:
### Just pushed some updates to the grant-extension-form:

Added a list of students to the component to support testing of the dropdown and placeholder behaviour. The default “Select a student” option has been updated accordingly.

Validation error messages are now shown for the student and reason fields when they're touched but not filled out, or when the "Grant Extension" button is clicked.

The Additional Notes field now indicates that it’s optional.

The "Grant Extension" button now disables on submit and resets all form fields after a successful submission.

Added inline comments to clarify the form logic and improve maintainability.

These updates reflect the requirements outlined in my assigned ticket, along with suggestions provided by Samindi.
Please note that this implementation is front-end only and there is no API integration at this stage. The form is still under development and will be enhanced in upcoming tasks.

SahiruWithanage commented:
Great work! I went through everything and didn’t run into any issues – it all worked smoothly. I also checked the console log, and everything seemed to be working fine there as well. 

<img width="1440" alt="Screenshot 2025-04-06 at 3 09 31 am" src="https://github.com/user-attachments/assets/a9aecd23-c6db-49d5-8dfb-a56093ae435f" />

<img width="1439" alt="Screenshot 2025-04-06 at 3 17 16 am" src="https://github.com/user-attachments/assets/89baa83f-4f20-4d80-9b75-6c8d34e38ac7" />

I do have a couple of suggestions though. One is to consider adding a word limit to the "Reason" and "Additional Notes" fields so the input doesn’t exceed a certain limit.

<img width="1440" alt="Screenshot 2025-04-06 at 3 20 07 am" src="https://github.com/user-attachments/assets/0185ba47-00b4-47d6-bc2a-683bc2a57851" />

Another idea is to include a “Cancel” or “Clear Form” button that resets everything back to the default state – that would be a nice touch for usability. Also, I think the suggestions made by Samindi were really great and worth considering as well. Overall, awesome job!

SahiruWithanage reviewed (changes requested):
Great work! I went through everything and didn’t run into any issues – it all worked smoothly. I also checked the console log, and everything seemed to be working fine there as well. 

<img width="1440" alt="Screenshot 2025-04-06 at 3 09 31 am" src="https://github.com/user-attachments/assets/a9aecd23-c6db-49d5-8dfb-a56093ae435f" />

<img width="1439" alt="Screenshot 2025-04-06 at 3 17 16 am" src="https://github.com/user-attachments/assets/89baa83f-4f20-4d80-9b75-6c8d34e38ac7" />

I do have a couple of suggestions though. One is to consider adding a word limit to the "Reason" and "Additional Notes" fields so the input doesn’t exceed a certain limit.

<img width="1440" alt="Screenshot 2025-04-06 at 3 20 07 am" src="https://github.com/user-attachments/assets/0185ba47-00b4-47d6-bc2a-683bc2a57851" />

Another idea is to include a “Cancel” or “Clear Form” button that resets everything back to the default state – that would be a nice touch for usability. Also, I think the suggestions made by Samindi were really great and worth considering as well. Overall, awesome job!

b0ink commented:
The current scaling of the form and text seems to be really small
![image](https://github.com/user-attachments/assets/2e108f95-6342-4c37-94b5-d950711049f9)

You should also be using [Material UI components](https://material.angular.io/components/form-field/overview) for elements like the select dropdown, slider, buttons, inputs, etc. This should reduce the amount of tailwind classes you're currently using

I'm not familiar with the guidelines of this project, but perhaps this could be a modal/dialogue pop up, similar to the extension request form component that students are able to submit. Are you working off any existing documentation that you could link here for reference?

JoeMacl commented:
> The current scaling of the form and text seems to be really small ![image](https://private-user-images.githubusercontent.com/40929320/430726627-2e108f95-6342-4c37-94b5-d950711049f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwODI5NTYsIm5iZiI6MTc0NDA4MjY1NiwicGF0aCI6Ii80MDkyOTMyMC80MzA3MjY2MjctMmUxMDhmOTUtNjM0Mi00YzM3LTk0YjUtZDk1MDcxMTA0OWY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDAzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmYzFlNGUzYWM1ZDhkNWQwMmM3MWYwYzg0MWMwY2ZlYTk0NjM3YTZjOGVjMTVlYWRjMzI3MTU4NWQ5MTI2NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-_Gs6ArmpEo9a2Dy5Ma7FL4Zh0Nf0r4jWtaw0DRrlVU)
> 
> You should also be using [Material UI components](https://material.angular.io/components/form-field/overview) for elements like the select dropdown, slider, buttons, inputs, etc. This should reduce the amount of tailwind classes you're currently using
> 
> I'm not familiar with the guidelines of this project, but perhaps this could be a modal/dialogue pop up, similar to the extension request form component that students are able to submit. Are you working off any existing documentation that you could link here for reference?

JoeMacl commented:
> The current scaling of the form and text seems to be really small ![image](https://private-user-images.githubusercontent.com/40929320/430726627-2e108f95-6342-4c37-94b5-d950711049f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwODI5NTYsIm5iZiI6MTc0NDA4MjY1NiwicGF0aCI6Ii80MDkyOTMyMC80MzA3MjY2MjctMmUxMDhmOTUtNjM0Mi00YzM3LTk0YjUtZDk1MDcxMTA0OWY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDAzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmYzFlNGUzYWM1ZDhkNWQwMmM3MWYwYzg0MWMwY2ZlYTk0NjM3YTZjOGVjMTVlYWRjMzI3MTU4NWQ5MTI2NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-_Gs6ArmpEo9a2Dy5Ma7FL4Zh0Nf0r4jWtaw0DRrlVU)
> 
> You should also be using [Material UI components](https://material.angular.io/components/form-field/overview) for elements like the select dropdown, slider, buttons, inputs, etc. This should reduce the amount of tailwind classes you're currently using
> 
> I'm not familiar with the guidelines of this project, but perhaps this could be a modal/dialogue pop up, similar to the extension request form component that students are able to submit. Are you working off any existing documentation that you could link here for reference?

Thanks for the feedback!
I'll start updating the form to use Angular Material as suggested. As for the modal/dialogue approach I went back through the documentation here --> https://thoth-tech.netlify.app/products/ontrack/projects/staff-grant-extension/design/ it doesn't specifically state it but implies to use the modal/dialog approach. I'll begin refactoring the form to follow that as well!

JoeMacl committed:
feat: update progress on grant extension form

JoeMacl commented:
### This PR implements the Grant Extension form as an Angular Material Dialog.

Please note: The form is accessible as a student for testing purposes. As there are issues preventing admins from accessing units on the 9.x branch. Once those issues are resolved, this will be updated so that only staff users can access the feature.

### Changes Made:
Grant Extension form implemented as a Material Dialog

Utilizes Material UI components (form fields, select dropdown, slider, textarea, buttons)

Form validation for required fields

Slider for extension duration (1-30 days)

Temporary visibility for students to allow testing due to current admin access issues

zara1504 commented:
After testing the grant extension component (in the student area, which I understand is here just for testing), I was able to get a console output that my extension has been submitted to the student I had selected and see the number of days and reasoning. Everything mentioned above that the form should include was there, with the required fields working and coming up with a message underneath when not filled out. Along with this, the slider works from 1-30 days. After seeing the comments written above, I can see that a cancel button has been added and works. 

I did see there was an empty file, grant-extension-form.component.scss. If it's not needed in this component, I would recommend removing it so it doesn't cause any clutter. 

Overall, everything worked for me during testing. Good job :).

SahiruWithanage reviewed (approved):
Nice work on the changes – everything looks good to me!

b0ink commented on src/assets/fonts/grotesk/grotesk-italic.otf:
Why was this font removed?

b0ink commented on src/assets/fonts/grotesk/grotesk-regular.otf:
Why was this font removed?

b0ink commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.html:
```diff
@@ -0,0 +1,67 @@
+    tickInterval="5"
+    thumbLabel
+    class="w-full"
+    style="width: 100%; max-width: 600px; min-width: 300px;"
```
move the inline styles as tailwind classes, here `w-full` and `width: 100%` achieve the same result

b0ink commented on src/app/admin/modals/grant-extension-form/grant-extension-form.component.ts:
```diff
@@ -0,0 +1,101 @@
+  isSubmitting = false;
+
+  // Temporary values will be replaced with dynamic context
+  unitId = 1;
+  taskDefinitionId = 25;
```
You could probably use @Input variables and then fake the data from wherever you display this modal, that way you could plug-n-play this modal pretty quickly once the api side is done

b0ink commented on src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.component.ts:
```diff
@@ -66,4 +73,16 @@ export class TaskStatusCardComponent implements OnChanges, AfterViewInit {
+
+  openGrantExtensionDialog(): void {
+    this.dialog.open(GrantExtensionFormComponent, {
+      width: '600px',
```
use max-width instead, and set width: to 100% to ensure better responsiveness on smaller screens

b0ink reviewed (commented):
Hi @JoeMacl 

I wasn't able to navigate to the grant extension route as it kept redirecting me back to the home page, but i have left a few notes regarding your changes - looks all good to me, just a few areas that you could change so that when the API routes are added this component can be merged relatively smoothly without additional changes

JoeMacl committed:
style: improve modal responsiveness and log form submission

aNebula commented:
~@JoeMacl  - why was there a need for a force push? Also, you need to get this reviewed again after any and all code changes. Please contact @b0ink  and get this task completed.~
Is this PR not valid any more, because I also see https://github.com/thoth-tech/doubtfire-web/pull/330 ?
In that case, @JoeMacl  please close the PR.

JoeMacl commented:
I'm closing this PR as [PR #330](https://github.com/thoth-tech/doubtfire-web/pull/330) includes a more updated and integrated implementation of the Staff Grant Extension frontend work.
